src/librustc_mir/interpret/place.rs

```diff
@@ -216,21 +216,40 @@ impl<'tcx, Tag> MPlaceTy<'tcx, Tag> {
             _ => bug!("vtable not supported on type {:?}", self.layout.ty),
         }
     }
+
+    #[inline(always)]
+    pub fn to_const_value(self, tcx: TyCtxt<'tcx>) -> ConstValue<'tcx> {
+        match self.mplace.ptr {
+            Scalar::Ptr(ptr) => {
+                let alloc = tcx.alloc_map.lock().unwrap_memory(ptr.alloc_id);
+                ConstValue::ByRef { alloc, offset: ptr.offset }
+            },
+            Scalar::Raw { data, .. } => {
+                assert_eq!(data, self.layout.align.abi.bytes().into());
+                ConstValue::Scalar(Scalar::zst())
```

Comment:
Why does this make any sense?

Reply:
If the mplace has an integer address and you read from it at compile time (which is what all users of this function want), it must be a ZST.

Reply:
Makes sense, but it'd be better if we could assert this. Can't we do `assert!(self.layout.zst())`?